### ISSUE TITLE
Mobile Safari gives different suggestions compared to other browsers

### ISSUE THREAD

**Issue by bluthen** [kwonoj/hunspell-asm#187]
I don't really understand why but on Mobile safari it offers more suggestions than on other browsers.

I thought it was because of the timeouts in hunspell, but if I change the header files and recompile, it didn't help. Also when I reduce the dictionary to a small number of words (like in the zip) it still does it.

The header file: src/hunspell/suggestmgr.hxx
TIMELIMIT define.

Maybe a bad expection, but I expect it to act the same. Not so sure where to start next, I guess I need to make sure I have the right maps and debug hunspell as it comes up with the suggestion list?

Attached is a minimal example:

```
unzip hunspellasm_test2.zip
cd hunspellasm_test2
npm install
npm run watch
```
In another terminal
```
cd hunspellasm_test2
http-server
```

Open up browser to where it is listening to:

Expect to show on all browsers including mobile safari:
`["Marshmallow"]`

On mobile safari you get:
`["Marshmallow","Marshall","Marshland","Marshal"]`

[hunspellasm_test2.zip](https://github.com/kwonoj/hunspell-asm/files/3278666/hunspellasm_test2.zip)

**Comment by kwonoj** [comment]
Huh, interesting. I would expect the result would be same across browsers too. For debugging, I think your step would be correct, with debug build of hunspell (and symbols) step through differences.

I don't own any ios devices so it's impossible to track it down myself, but you can build wasm binary for debug build via https://github.com/kwonoj/docker-hunspell-wasm.

**Comment by bluthen** [comment]
I've found this pretty difficult, but will keep trying.

So far I've, modified the Dockerfile to include CFLAGS=' -g4 -O2 ' and CXXFLAGS=' -g4 -O2 ' for emconfigure and emmake. Then modify build/build.sh to include CFLAGS=' -g4 -O2 ' and CXXFLAGS=' -g4 -O2 ' as arguments to em++.

In the debugger, I see running as wasm-0029c2ca. I add sourcemap  hunspell.wasm.map. Several files show up hunspell.cxx, suggestmgr.cxx, etc, but they are all empty inside.

**Comment by bluthen** [comment]
So far I've isolated the issue to this:

suggestmgr.cxx around line 1340

```
  int is_swap = 0;
  int re = 0;
  double fact = 1.0;
  printf("!!!1 fact=%f\n", fact);
  if (pAMgr) { 
    printf("!!!2 fact=%f\n", fact);
    int maxd = pAMgr->get_maxdiff();
    printf("!!!3 fact=%f, %d\n", fact, maxd);
    if (maxd >= 0) 
      fact = (10.0 - maxd) / 5.0;
    printf("!!!4 fact=%f, %d\n", fact, maxd);
  }
```

On non-mobile safari it prints out:
```
!!!1 fact=1.000000
!!!2 fact=1.000000
!!!3 fact=1.000000, -1
!!!4 fact=1.000000, -1
```

On mobile safari it prints out:
```
!!!1 fact=1.000000
!!!2 fact=1.000000
!!!3 fact=1.000000, -1
!!!4 fact=0.000000, -1
```

Pretty strange. The variable fact just gets set to 0, for no reason that I could tell.

**Comment by kwonoj** [comment]
huh weird, `(10.0 - maxd) / 5.0` seems harmless enough across any env. I have to admit I have no idea what could cause this. Speculative guess maybe some build config in hunspell + emscripten causes this issues? curious if this can be reproducible out of hunspell's codebase.

**Comment by bluthen** [comment]
So far this has been the only thing that has solved it. Without the printf in the else it still is zero. I don't consider this a good solution as it makes no sense.


[hunspell-asm_issue187.patch.txt](https://github.com/kwonoj/hunspell-asm/files/3311390/hunspell-asm_issue187.patch.txt)

**Comment by kwonoj** [comment]
> Without the printf in the else it still is zero.

That gives me suspicion there are potential memory alignment or similar in certain cases. and yes, that patch doesn't sounds ok too.

**Comment by bluthen** [comment]
Compiling with `-O0` seems to get expected results. Both `-O1` and `-O2` express the same problem.

**Comment by kwonoj** [comment]
huh (2), optimization plays odd here? that explains some. I can bump up version with optimization turned off, expect perf / size wouldn't too hurt with it. Will bring up PR soon.

**Comment by kwonoj** [comment]
well, seems size is bit hurting 😢 created PR https://github.com/kwonoj/hunspell-asm/pull/192, @bluthen if you can try this PR and if it works fine, I'll check this in and publish new version.

**Comment by bluthen** [comment]
Yeah over double size.
I tried the PR. Suggestions came in as expected.

**Comment by kwonoj** [comment]
Personally I don't care that much about bytesize due to my running environment, but probably it'll be bit some pain for others. For now, I'll follow correct behavior over size gain - if I can get some light from emscripten issues maybe I can bring optimization back.

**Comment by kwonoj** [comment]
https://github.com/kwonoj/hunspell-asm/releases/tag/v3.0.2 is out now.

**Comment by bluthen** [comment]
I wonder if it could be worth it to dynamically load the optimized or non-optimized webasm based on if iOS or not. This would further increase the bundle size though, or need to make it an external file.

**Comment by kwonoj** [comment]
> dynamically load the optimized or non-optimized webasm based on if iOS or not.

This'll defeat bundler's tree shaking, ends up pretty much all platform will load both binary - it'll be quite loss so probably I won't pursue it.

Honestly atm, I'd expect safari should fix bug, otherwise you may use 2.x for old version (or known non-working version) of safari instead. That was reason I did semver major for enabling optimization again.

**Comment by bluthen** [comment]
The issue seems resolved in iOS/iPadOS 13.1 but not 12.4.1 (latest 12)

Several devices can not upgrade to 13, so will use unoptimized version until those are used less or figure out my own dynamic network download runtime loading.

This is a great package. Thank you!
